In [14]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
%matplotlib inline

In [53]:
sample = np.array([float(line.rstrip('\n')) for line in open('9_2_Malyshev.txt', 'r')])

Т.к. $X_i = \beta_1 + i\beta_2 + \varepsilon_0 + \ldots + \varepsilon_i$, то
$$X_0 = \beta_1 + \varepsilon_0$$
$$X_1 - X_0 = \beta_2 + \varepsilon_1$$
$$\ldots$$
$$X_n - X_{n-1} = \beta_2 + \varepsilon_n$$

Получаем линейно-регрессионную модель для $\vec{Y} = (X_0, X_1 - X_0, \ldots, X_n - X_{n-1})^{T}$
$$\vec{Y} = Z(\beta_1, \beta_2)^{T} + \vec{\epsilon}$$

In [54]:
Y = np.array([sample[0]] + [sample[i] - sample[i-1] for i in range(1, len(sample))])

In [59]:
n = len(sample)
Z = np.array([[1, 0]] + [[0, 1] for i in range(n-1)])

Учитывая, что оценка $(\beta_1, \beta_2)$ равна $$(\hat{\beta}_1, \hat{\beta}_2)^{T} = (Z^{T}Z)^{-1}Z^{T}\vec{Y}$$

In [60]:
def estimation_of_bettas(Y, Z):
    return np.dot(np.dot(np.linalg.inv(np.dot(Z.transpose(),Z)),Z.transpose()),Y)  # dot - операция умножения матриц

In [75]:
betas_est = estimation_of_bettas(Y, Z)
print(betas_est)  # оценка \beta_1, \beta_2

[ 626.348568      6.05324828]


Несмещенная оценка $\sigma^2$ равна$$\hat{\sigma}^2 = \frac{1}{n-1}|Y - Z\hat\Theta|^2 = \frac{1}{n-1}\sum(Y_i - \overline Y)^2$$

In [70]:
def sigma_sqrt_estimation(Y):
    n = len(Y[1:])  # т.к. мы берем \epsilon_i, где i ≥ 1
    avg = np.average(Y[1:])
    return 1/(n-1) * np.sum((Y[1:] - avg)**2)

In [79]:
sigma_sqrt = sigma_sqrt_estimation(Y)  #оценка \sigma^2
print(sigma_sqrt)

0.04673124236


Из соотношения $\varepsilon_i = \varepsilon_i^t \beta_2 $ оценка $\sigma^2_{t}$ равна $$\hat{\sigma}^2_{t} = \frac{\sigma^2}{\beta^2_2}$$

In [76]:
sigma_sqrt_t = sigma_sqrt / betas_est[1]

In [77]:
print(sigma_sqrt_t)

0.00772002735931


#### Вывод.
Оценка начального растояния равна
$$\hat{\beta}_1 = 626.348568$$
скорость равна
$$\hat{\beta}_2 = 6.05324828$$
дисперсия ошибки равна
$$\hat{\sigma^2} = 0.04673124236,$$
а оценка дисперсии отчета времени равна
$$\hat{\sigma^2_t} = 0.00772002735931.$$